compiler/resolution/functionResolution.cpp

```diff
@@ -5184,81 +5194,101 @@ resolveNew(CallExpr* call)
+
         } else {
-          toReplace->replace(new UnresolvedSymExpr(ct->defaultInitializer->name));
+          FnSymbol* ctInit = ct->defaultInitializer;
+
+          toReplace->replace(new UnresolvedSymExpr(ctInit->name));
         }
       }
     }
 
     // 2: move the 1st argument into the baseExpr
     // 3: make it a normal call and not a PRIM_NEW
     Expr* arg = call->get(1);
+
     arg->remove();
+
     call->primitive = NULL;
-    call->baseExpr = arg;
+    call->baseExpr  = arg;
+
     parent_insert_help(call, call->baseExpr);
 
     if (initCall) {
       // 4: insert the allocation for the instance and pass that in as an
       // argument if we're making a call to an initializer
-      Type* typeToNew = toReplace->symbol()->typeInfo();
-      VarSymbol* new_temp = newTemp("new_temp", typeToNew);
+      Type*      typeToNew = toReplace->symbol()->typeInfo();
+      VarSymbol* newTmp    = newTemp("new_temp", typeToNew);
+
+      VarSymbol* newMT     = newTemp("_mt",      dtMethodToken);
+
       if (typeToNew->symbol->hasFlag(FLAG_GENERIC)) {
-        USR_FATAL(call, "Sorry, new style initializers don't work with generics yet.  Stay tuned!");
+        USR_FATAL(call,
+                  "Sorry, new style initializers don't work with "
+                  "generics yet.  Stay tuned!");
       }
-      DefExpr* def = new DefExpr(new_temp);
-      Expr* insertBeforeMe = ((!isBlockStmt(call->parentExpr)) ?
-                              call->parentExpr : call);
+
+      DefExpr* def            = new DefExpr(newTmp);
+      Expr*    insertBeforeMe = NULL;
+
+      if (isBlockStmt(call->parentExpr) == true) {
+        insertBeforeMe = call;
+
+      } else {
+        insertBeforeMe = call->parentExpr;
+      }
+
       insertBeforeMe->insertBefore(def);
+
       resolveExpr(def);
 
       if (!isRecord(typeToNew) && !isUnion(typeToNew)) {
-        BlockStmt* allocCallBlock = new BlockStmt();
+        BlockStmt* allocBlock = new BlockStmt();
+        CallExpr*  allocCall  = callChplHereAlloc(typeToNew);
+
+        insertBeforeMe->insertBefore(allocBlock);
 
-        CallExpr* innerCall = callChplHereAlloc(typeToNew);
-        CallExpr* allocCall = new CallExpr(PRIM_MOVE, new_temp,
-                                           innerCall);
-        insertBeforeMe->insertBefore(allocCallBlock);
-        allocCallBlock->insertAtTail(allocCall);
+        allocBlock->insertAtTail(new CallExpr(PRIM_MOVE,   newTmp, allocCall));
+        allocBlock->insertAtTail(new CallExpr(PRIM_SETCID, newTmp));
 
-        CallExpr* setIdCall = new CallExpr(PRIM_SETCID, new_temp);
-        allocCallBlock->insertAtTail(setIdCall);
-        normalize(allocCallBlock);
-        resolveBlockStmt(allocCallBlock);
+        normalize(allocBlock);
+
+        resolveBlockStmt(allocBlock);
       }
-      call->insertAtTail(new NamedExpr("meme", new SymExpr(new_temp)));
+
+      call->insertAtHead(new SymExpr(newTmp));
+      call->insertAtHead(new SymExpr(newMT));
```

Comment:
Since this actually impacts functionality, I will not be upset by dealing with it on my branch.